src/main/java/world/bentobox/parkour/listeners/CourseRunnerListener.java

```diff
@@ -71,11 +72,40 @@ public void onVisitorArrive(IslandEnterEvent e) {
         } else if (!parkourRunManager.timers().containsKey(e.getPlayerUUID())) {
             user.notify("parkour.to-start");
         }
-        if (island.getFlag(addon.PARKOUR_CREATIVE) <= island.getRank(user)) {
-            user.setGameMode(GameMode.CREATIVE);
-        } else {
-            user.setGameMode(GameMode.SURVIVAL);
+
+        if (addon.inWorld(user.getLocation())) {
+            // only affect gamemode if they are currently in the parkour world
+            // cross world teleportation is handled in onPlayerChangeWorld
+            if (island.getFlag(addon.PARKOUR_CREATIVE) <= island.getRank(user)) {
+                user.setGameMode(GameMode.CREATIVE);
+            } else {
+                user.setGameMode(GameMode.SURVIVAL);
+            }
         }
+
+
+    }
+
+    @EventHandler
+    public void onPlayerChangeWorld(PlayerChangedWorldEvent event) {
+        User user = User.getInstance(event.getPlayer());
+        Location currentLocation = user.getLocation();
+        if (!addon.inWorld(currentLocation)) {
+            return;
+        }
+        // Check if player is in a parkour course
+        Optional<Island> course = addon.getIslandsManager().getIslandAt(currentLocation);
+        if (course.isEmpty()) {
```

Comment:
If you use isPresent, then you don't need to check isEmpty or make the local variable "course".

Reply:
done